### ISSUE TITLE
Fine tune TROCR on french medical data

### ISSUE THREAD

**Issue by Serge9744** [NielsRogge/Transformers-Tutorials#270]
Hi @NielsRogge ,

I am currently trying to fine tune TrOCR on handwritten French medical data that we labelled from images of handwritten texts.

So what I did was :
1/ Loading the first stage TROCR model 
2/ Create a BPE tokenizer on the training French medical data and then assing it for the decoder

# this is the custom tokenizer :
   loaded_tok = Tokenizer.from_file(text_processor_path)

# then i wrap it :
    wrapped_tokenizer = PreTrainedTokenizerFast(
        tokenizer_object=loaded_tok,
        bos_token= '<s>',
        eos_token= '</s>',
        pad_token='[PAD]',
        mask_token= '<mask>',
        unk_token= '<unk>',
        sep_token='</s>',
        cls_token= '<s>')
    
    #creating train and dev tensors
    train_dataset= HandwrittenDataSet(df_train_total,processor , wrapped_tokenizer , max_pad_length)
    dev_dataset= HandwrittenDataSet(df_dev,processor, wrapped_tokenizer , max_pad_length) 
    #creating train and dev loaders
    train_dataloader = DataLoader(train_dataset, batch_size=batch_size, shuffle=True)
    dev_dataloader = DataLoader(dev_dataset, batch_size=batch_size)
    
# Here I load ../models/TrOCR_large_stage1
    device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
    base_model = VisionEncoderDecoderModel.from_pretrained("../models/TrOCR_large_stage1")
    base_model.to(device)
    
   # here i assign the decoder tokens  and vocab size to be the one from the custom decoder

    base_model.config.decoder_start_token_id = wrapped_tokenizer.bos_token_id
    base_model.config.pad_token_id = wrapped_tokenizer.pad_token_id
    base_model.config.vocab_size = wrapped_tokenizer.vocab_size
    base_model.config.eos_token_id = wrapped_tokenizer.sep_token_id
    base_model.tokenizer = wrapped_tokenizer

My questions would be : 

1/ Is it preferable , instead of starting from the TROCR large stage 1, to start from a VIT Model (vit base from google) and Camembert as the decoder ?Or my approach to use the TROCR large stage 1 + a custom BPE tokenizer would be fine on French handrwitten data?

2/There are some typos on the texts , for example we have in an image a text like this : "Dibète" and the true word  is "Diabète". Hence the label is uncorrect and missing an "a" . 
Should we correct the label to "Diabète" to start learning even if the character is missing? As there is a language model and beam search at the decoder stage , could it kind of learn an auto correction and predict Diabètes even if it sees "Dibètes"? 
Or should we keep the labels as they are for better learning and then correct them at the prediction stage?

**Comment by Serge9744** [comment]
Hi !
Many thanks for the reply !
Should we correct the labels or let it like with errors ? Woukd you have an idea on that topic ?
Thanks

**Comment by NielsRogge** [comment]
Sorry also had to answer your second question :) I would make the model learn the correct spelling, so correct the labels

**Comment by Serge9744** [comment]
Thanks, I understood your argument for the trocr stage 1 model, but what is your idea on the label correction ? The correct conditional probabilities would be learnt ? Like an embedded auto correct ? 

Also Should I train a BPE tokenizer like I did on the training data or use the decoder tokenizer of Camembert ?

Thankd for everything

**Comment by NielsRogge** [comment]
> The correct conditional probabilities would be learnt ? Like an embedded auto correct ?

Yes the model should learn to output the correct text.

You can use the tokenizer of CamemBERT. You can create a model like so (for example):

```
from transformers import VisionEncoderDecoderModel

model = VisionEncoderDecoderModel.from_encoder_decoder_pretrained(
    "microsoft/swin-base-patch4-window7-224-in22k", "camembert-base"
)
```

**Comment by Serge9744** [comment]
Many thanks everything is clear then.

I'll use the tokenizer from Camembert instead of a custom BPE learnt on the text then

**Comment by Serge9744** [comment]
Hi @NielsRogge ,

I have run a few experiments, playing with the different parameters and looking at the documentation.

What I did was the following:

I have ~2000 images of written text . What I do is that I cut the data into train dev tests .

 I augment the train set doing darkening, brigthening , upsizing, downsizing , addind noise etc.. only . 
This gives me 13 000 images roughly
Then I do a pre processing where I use contrast and sharpening.
You can see one exemple (Coviran (Tension)) with the output of a trained model.
Basically my intuition is that the cleaner the data is the better it is going to learn , and I'll use the same process in prod for prediction of course
 
![coviran](https://user-images.githubusercontent.c

So after I do that I am ready to train the model . I use:

class HandwrittenDataSet(Dataset):
    
    def __init__(self,df,processor,text_tokenizer,max_pad_length,file_name_col , text_col):
        self.df= df
        self.processor = processor
        self.max_pad_length = max_pad_length
        self.text_tokenizer = text_tokenizer
        self.file_name_col =file_name_col 
        self.text_col = text_col
        
    def __len__(self):
        return len(self.df)
    
    def __getitem__(self,idx):
        image_name = self.df.loc[idx,self.file_name_col]
        handwritten_txt = self.df.loc[idx,self.text_col]
        image = Image.open(image_name)
        if len(np.array(image).shape)<3:
            image = cv2.cvtColor(np.array(image),cv2.COLOR_GRAY2RGB)
        pixel_values = self.processor(image , return_tensors = "pt").pixel_values
        labels = self.text_tokenizer(handwritten_txt,padding = "max_length" ,max_length= self.max_pad_length,truncation = True).input_ids
        labels = [label if label != self.text_tokenizer.pad_token_id else -100 for label in labels]
        encoding = {"pixel_values": pixel_values.squeeze(), "labels": torch.tensor(labels)}
        return encoding


And : 

loaded_tok = CamembertTokenizer.from_pretrained("../models/Camembert")

#creating train and dev tensors
train_dataset= HandwrittenDataSet(df_train_total,processor , loaded_tok , max_pad_length)
dev_dataset= HandwrittenDataSet(df_dev,processor, loaded_tok , max_pad_length) 
#creating train and dev loaders
train_dataloader = DataLoader(train_dataset, batch_size=batch_size, shuffle=True)
dev_dataloader = DataLoader(dev_dataset, batch_size=batch_size)

# I use Swin for Encoding and Camembert as decoding
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
base_model = VisionEncoderDecoderModel.from_encoder_decoder_pretrained("../models/Swin" , "../models/Camembert")

base_model.to(device)

## I use the tokens of Camembert
base_model.config.decoder_start_token_id = loaded_tok.bos_token_id
base_model.config.pad_token_id = loaded_tok.pad_token_id
# make sure vocab size is set correctly

# Vocab of camembert
base_model.config.vocab_size = loaded_tok.vocab_size

# set beam search parameters
base_model.config.eos_token_id = loaded_tok.sep_token_id
base_model.config.max_length = 64
base_model.config.early_stopping = True
base_model.config.no_repeat_ngram_size = 2
base_model.config.length_penalty = 1
base_model.config.num_beams =4

# tokenizer
base_model.tokenizer = loaded_tok 

#retrieve number of epochs

epoch_number = 30

batch_size = 8

optimizer = AdamW(base_model.parameters(), lr=5e-5)

Then I start training , actually even with all the random seeds set to 0 the model is not reproducible:

random.seed(0)
np.random.seed(0)
torch.manual_seed(0)
torch.cuda.manual_seed(0)
torch.backends.cudnn.deterministic = True
torch.backends.cudnn.benchmark = False
val_loss = []
base_model.zero_grad()
for epoch in range(epoch_number):  # loop over the dataset multiple times
# train
    base_model.train()
    train_loss = 0.0
    for batch in tqdm(train_dataloader):
        # get the inputs
        for k,v in batch.items():
            batch[k] = v.to(device)
        # forward + backward + optimize
        outputs = base_model(**batch)
        loss = outputs.loss
        loss.backward()
        optimizer.step()
        optimizer.zero_grad()
        train_loss += loss.item()
        
    #comment early stopping as the infrastructure is really unstable I do not want to overload calculations
    print(f'epoch {epoch} finished now evaluating loss')
    with torch.no_grad(): 
        valid_loss = predict_all_set(base_model , dev_dataloader)
        print('train loss : ' , train_loss/(len(train_dataloader)))
        print('val loss : ' , valid_loss)

Then I get a 4.8 CER on dev set and you can see some exemples of output on the dev set which I use to search for hyper params manually :

![exemples_medic](https://user-images.githubusercontent.com/43743992/233456672-21e72af3-3997-4503-b48f-0ae8912f649f.png)
om/43743992/233456485-e5638d46-a5f5-4a95-a5c1-88a74c3b4f4a.png)

==>Could you please help me and tell me what have I missed? beam_max_length affectation?
How can I get more reliable results?

==>How to get reproducible results? torch.manual_seed(0) does not allow reproducible 

Many many thanks for you help

**Comment by NielsRogge** [comment]
Can you please first fix the formatting of your message? It's pretty hard to read
